tests/late_joiners.py

```diff
@@ -0,0 +1,216 @@
+            check = infra.checker.Checker()
+
+            clients = []
+            with contextlib.ExitStack() as es:
+                LOG.info("Write messages to nodes using round robin")
+                clients.append(es.enter_context(first_node.user_client(format="json")))
+                for backup in backups:
+                    clients.append(es.enter_context(backup.user_client(format="json")))
+                node_id = 0
+                for id in range(1, TOTAL_REQUESTS):
+                    if id == 1:
+                        LOG.info(
+                            "Adding another node after f = 0 but before we need to send append entries"
+                        )
+                        # check that a new node can catch up naturally
+                        new_node = network.create_and_trust_node(
+                            lib_name=args.package, host="localhost", args=args,
+                        )
+                        assert new_node
+                        nodes_to_keep.append(new_node)
+                    node_id += 1
+                    c = clients[node_id % len(clients)]
+                    try:
+                        resp = c.rpc("LOG_record", {"id": id, "msg": long_msg})
+                    except Exception:
+                        LOG.info("Trying to access a suspended node")
+                    try:
+                        cur_primary, cur_term = network.find_primary()
+                        term_info[cur_term] = cur_primary.node_id
+                    except Exception:
+                        LOG.info("Trying to access a suspended node")
+                    id += 1
+
+                # wait for the last request to commit
+                first_catchup_msg = "Hello world!"
+                check_commit(
+                    c.rpc("LOG_record", {"id": 1000, "msg": first_catchup_msg}),
+                    result=True,
+                )
+                check(
+                    c.rpc("LOG_get", {"id": 1000}), result={"msg": first_catchup_msg},
+                )
+
+                # check that new node has caught up ok
+                with new_node.user_client(format="json") as c:
+                    check(
+                        c.rpc("LOG_get", {"id": 1000}),
+                        result={"msg": first_catchup_msg},
+                    )
+                # add new node to backups list
+                backups.append(new_node)
+
+                # check that a new node can catch up after all the requests
+                last_node = network.create_and_trust_node(
+                    lib_name=args.package, host="localhost", args=args,
+                )
+                assert last_node
+                nodes_to_keep.append(last_node)
+
+            ## send more messages I want to check that the new node will catchup and then start processing pre prepares, etc
+            clients = []
+            with contextlib.ExitStack() as es:
```

Comment:
This is roughly the same code as earlier. Could this be made into a function?

Reply:
Yes :)